test/common/grpc/google_async_client_cache_test.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/grpc/google_async_client_cache.h"
+
+#include "test/mocks/server/factory_context.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Grpc {
+namespace {
+
+class AsyncClientCacheTest : public testing::Test {
+public:
+  AsyncClientCacheTest() {
+    client_cache_singleton_ = std::make_unique<AsyncClientCacheSingleton>();
+  }
+
+  void expectClientCreation() {
+    factory_ = new Grpc::MockAsyncClientFactory;
+    async_client_ = new Grpc::MockAsyncClient;
+    EXPECT_CALL(async_client_manager_, factoryForGrpcService(_, _, true))
+        .WillOnce(Invoke([this](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool) {
+          EXPECT_CALL(*factory_, create()).WillOnce(Invoke([this] {
+            return Grpc::RawAsyncClientPtr{async_client_};
+          }));
+          return Grpc::AsyncClientFactoryPtr{factory_};
+        }));
+  }
+
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  Grpc::MockAsyncClientManager async_client_manager_;
+  Grpc::MockAsyncClient* async_client_ = nullptr;
+  Grpc::MockAsyncClientFactory* factory_ = nullptr;
+  NiceMock<Stats::MockIsolatedStatsStore> scope_;
+  std::unique_ptr<AsyncClientCacheSingleton> client_cache_singleton_;
+};
+
+TEST_F(AsyncClientCacheTest, Deduplication) {
+  Stats::IsolatedStoreImpl scope;
+  testing::InSequence s;
+
+  ::envoy::config::core::v3::GrpcService config;
+  config.mutable_google_grpc()->set_target_uri("dns://test01");
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential01");
+
+  expectClientCreation();
+  Grpc::RawAsyncClientSharedPtr test_client_01 =
+      client_cache_singleton_
+          ->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config)
+          ->getAsyncClient();
+  // Fetches the existing client.
+  EXPECT_EQ(test_client_01,
+            client_cache_singleton_
+                ->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config)
+                ->getAsyncClient());
+
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential02");
```

Comment:
Due to the chain of calls, I'm not sure here whether you are proving that you get a new client-cache after these credentials are updated, or whether you are proving that with the same client-cache you get new clients. I think it's the latter, but explicit temps and more comparisons would make it clearer.

Reply:
I think the logic looks much cleaner when they are flat. Since there are some EQs and some NEs, spending some more lines to abstract functions doesn't save much.

Reply:
I can accept that this is a judgement call, and I'm fine going with your judgement. However I don't think this addresses the point I raise in this comment. I don't know whether we are testing that we are rebuilding the client-cache or rebuilding the async-client, because in your test code you are doing both of those on every EXPECT.

Reply:
ah, I think I understand your concern. I'm testing the EQ of the RawAsyncClientSharedPtr created by two different series of calls (getOrCreateAsyncClientCache, getAsyncClient). It's no matter how many call we use to fetch the RawAsyncClientSharedPtr for the same config, the final object we get are the same. (Please correct me if I still misunderstand :> )

Reply:
Not exactly: it's more critical when you get a *different* result and I think the test should be clear about what caused the difference. Specifically, You are doing a()->b()->c() several times. To me there are two issues:
 - I am a human common-subsexpression-eliminator and subscribe to DRY (Don't Repeat Yourself) style, but it's just a preference. Usually that results in more compact code but it's not the most compelling reason for me to suggest the change.

More importantly:
 - if you are checking that a property of a()->b()->c() is "foo" and then you mutate state, and call a()->b()->c() and check that the property is now "bar", I don't know whether that's because a() reads the mutated state or b(). Applying DRY makes that explicit.

Reply:
Makes sense. I made some changes. Hope I understand correctly this time :)